### PR TITLE
unorderedmap: missing `space` instance for erasure and rehash ops

### PR DESCRIPTION
**Pull request by romintomasetti**
This PR is extracted from #6584 as requested by @dalg24.

It adds a `space` instance argument to map erasure and rehash operators for `Kokkos::UnorderedMap`.

~~It also makes `invalid_index` public, otherwise it cannot be used by client code, though the `invalid_index` is used *e.g.* by `Kokkos::UnorderedMap::find`.~~ Topic moved to #6641.

### CODE REVIEW

**Comment by romintomasetti** [comment]
@dalg24 @fnrizzi @masterleinad Would you have a minute to look at this one? Thanks!

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp [hunk 1]
```diff
@@ -223,16 +222,14 @@ struct UnorderedMapPrint {
 
   map_type m_map;
 
-  UnorderedMapPrint(map_type const& map) : m_map(map) {}
```
Comment (masterleinad) [inline containers/src/impl/Kokkos_UnorderedMap_impl.hpp:226]: Why did you change?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 2]
```diff
@@ -243,16 +243,19 @@ class UnorderedMap {
...
-  static const bool is_insertable_map =
+  static constexpr bool is_insertable_map =
       !has_const_key && (is_set || !has_const_value);
-  static const bool is_modifiable_map = has_const_key && !has_const_value;
-  static const bool is_const_map      = has_const_key && has_const_value;
+  static constexpr bool is_modifiable_map = has_const_key && !has_const_value;
+  static constexpr bool is_const_map      = has_const_key && has_const_value;
+
+  static constexpr size_type invalid_index =
+      KOKKOS_INVALID_INDEX_TYPE(size_type);
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:258]: Why is it important for `invalid_index` to be public? How would it be used in user code?

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp [hunk 3]
```diff
@@ -76,19 +75,19 @@ struct UnorderedMapErase {
   using size_type       = typename map_type::size_type;
   using key_type        = typename map_type::key_type;
   using value_type      = typename map_type::impl_value_type;
+  using policy_type =
+      Kokkos::RangePolicy<execution_space, IndexType<size_type>>;
 
   map_type m_map;
 
-  UnorderedMapErase(map_type const& map) : m_map(map) {}
```
Comment (masterleinad) [inline containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: Why did you remove?

Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: 1. `UnorderedMapErase` is an aggregate and as such can be initialized using aggregate initialization.
2. In my opinion, the constructor was not correct in the sense that, bearing in mind that the map is a "shared pointer to some memory", it should have been `UnorderedMapErase(map_type map) : m_map(std::move(map)) {}` to clearly show that shared ownership is requested.

All in all I decided to remove the constructor :pencil2:

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 2 above)
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp:258]: In one of my tests, I do:
```c++
using map_t = Kokkos::UnorderedMap<int, double, execution_space>;

...

map_t map(...);

...

Kokkos::parallel_for(
    ...,
    KOKKOS_LAMBDA(...) {

        ....

        if(const auto direct_index = map.find(my_key); direct_index != ~static_cast<typename umap_t::size_type>(0))
        {
            <use `direct_index` here>
        }
    }
);

...
```

So in this case it would be nice to have access to `invalid_index`, which in fact is a value that can be returned by `find`, which is **a public member function**. IMO we don't even need a use case to decide if `invalid_index` should be public. I think it should because it can be returned by a public member function.

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp (hunk 1 above)
Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:226]: For the same reason as for `UnorderedMapErasure`, see my comment https://github.com/kokkos/kokkos/pull/6610#discussion_r1402100981 #5980

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp (hunk 3 above)
Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: @masterleinad What's your opinion on this then ? :)

Comment (masterleinad) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: My preference is to minimize drive-by changes. 🙂

Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: @masterleinad I already extracted these changes from #6584... I'm not against making another "extracted from" PR with the "remove useless constructors" change if you really want to and in general I agree with the "one PR per topic without drive-by change", but it makes the contribution process very lengthy for such a "deletion of useless line of code change" case... Especially that in this PR I already modify the `UnorderedMapErase` & Co. functors, so deleting their useless constructor can arguably not be a drive-by change :angel:

What do you think?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 4]
```diff
@@ -465,26 +476,24 @@ class UnorderedMap {
...
+  bool end_erase(const execution_space &space = execution_space{}) {
     bool result = erasable();
     if (is_insertable_map && result) {
-      execution_space().fence(
-          "Kokkos::UnorderedMap::end_erase: fence before erasing");
-      Impl::UnorderedMapErase<declared_map_type> f(*this);
-      f.apply();
-      execution_space().fence(
-          "Kokkos::UnorderedMap::end_erase: fence after erasing");
+      space.fence("Kokkos::UnorderedMap::end_erase: fence before erasing");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:493]: Same here.

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp [hunk 5]
```diff
@@ -47,16 +47,15 @@ struct UnorderedMapRehash {
...
   map_type m_dst;
   const_map_type m_src;
 
-  UnorderedMapRehash(map_type const& dst, const_map_type const& src)
-      : m_dst(dst), m_src(src) {}
-
-  void apply() const {
-    parallel_for("Kokkos::Impl::UnorderedMapRehash::apply", m_src.capacity(),
-                 *this);
+  void apply(const execution_space& space = execution_space{}) const {
```
Comment (masterleinad) [inline containers/src/impl/Kokkos_UnorderedMap_impl.hpp:56]: Is there a reason to keep the default argument? Since this is not user-facing, I would rather add passing an execution space instance on the call site.

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp [hunk 6]
```diff
@@ -76,19 +75,19 @@ struct UnorderedMapErase {
...
   using value_type      = typename map_type::impl_value_type;
+  using policy_type =
+      Kokkos::RangePolicy<execution_space, IndexType<size_type>>;
 
   map_type m_map;
 
-  UnorderedMapErase(map_type const& map) : m_map(map) {}
-
-  void apply() const {
+  void apply(const execution_space& space = execution_space{}) const {
```
Comment (masterleinad) [inline containers/src/impl/Kokkos_UnorderedMap_impl.hpp:83]: I would similarly avoid the default argument here.

File: core/src/Kokkos_Macros.hpp [hunk 7]
```diff
@@ -532,7 +532,11 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #define KOKKOS_ENABLE_CUDA_LDG_INTRINSIC
 #endif
 
-#define KOKKOS_INVALID_INDEX (~std::size_t(0))
+//! Invalid index of a given @p __type__.
+#define KOKKOS_INVALID_INDEX_TYPE(__type__) ~static_cast<__type__>(0)
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:536]: If we are to provide this macro at all, this should be prefixed with `KOKKOS_IMPL`.

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp (hunk 3 above)
Comment (masterleinad) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:82]: You likely get a pull request merged if you change fewer lines of code but it's not a showstopper for me.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 8]
```diff
@@ -465,26 +476,24 @@ class UnorderedMap {
     return is_insertable_map ? get_flag(erasable_idx) : false;
   }
 
-  bool begin_erase() {
+  bool begin_erase(const execution_space &space = execution_space{}) {
     bool result = !erasable();
     if (is_insertable_map && result) {
-      execution_space().fence(
+      space.fence(
```
[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp:482]: Ok, I'll remove the fence. Should I provide an overload that does not expect an execution space instance?

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 4 above)
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp:493]: Ok, I'll remove the fences. Should I provide an overload that does not expect an execution space instance?

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp (hunk 5 above)
Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:56]: I agree, I removed the default.

File: containers/src/impl/Kokkos_UnorderedMap_impl.hpp (hunk 6 above)
Comment (romintomasetti) [reply containers/src/impl/Kokkos_UnorderedMap_impl.hpp:83]: Indeed, consider it done :wink:

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp]: BTW, `begin_erase` calls `set_flag`. And `set_flag` makes a `Kokkos::fence(..)`.

I guess I should create an issue for that ?

[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp]: Sure.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp]: Sure, for the issue, the overload or both ? :smile:

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 9]
```diff
@@ -413,23 +414,33 @@ class UnorderedMap {
   /// This is <i>not</i> a device function; it may <i>not</i> be
   /// called in a parallel kernel.
   bool rehash(size_type requested_capacity = 0) {
+    return rehash(execution_space{}, requested_capacity);
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:417]: We should probably fence after calling the overload taking an execution space instance.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 10]
```diff
@@ -413,23 +414,33 @@ class UnorderedMap {
...
+  }
+
+  bool rehash(const execution_space &space, size_type requested_capacity = 0) {
     const bool bounded_insert = (capacity() == 0) || (size() == 0u);
-    return rehash(requested_capacity, bounded_insert);
+    return rehash(space, requested_capacity, bounded_insert);
   }
 
   bool rehash(size_type requested_capacity, bool bounded_insert) {
+    return rehash(execution_space{}, requested_capacity, bounded_insert);
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:426]: Here as well.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 11]
```diff
@@ -465,26 +476,19 @@ class UnorderedMap {
     return is_insertable_map ? get_flag(erasable_idx) : false;
   }
 
-  bool begin_erase() {
+  bool begin_erase(const execution_space &space) {
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:479]: We still need the overload without execution space instance for backward-compatibility.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp:417]: Do you agree that it's a `Kokkos::fence` that's needed?
```c++
bool rehash(size_type requested_capacity = 0) {
  bool result = rehash(execution_space{}, requested_capacity);
  Kokkos::fence("Kokkos::UnorderedMap::rehash: fence after rehasing");
  return result;
}
```

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 11 above)
Comment (romintomasetti) [reply containers/src/Kokkos_UnorderedMap.hpp:486]: Ok, I've written one. But I'm not sure it is done in the correct way. Could you check please?

File: core/src/Kokkos_Macros.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/src/Kokkos_Macros.hpp:536]: If you agree, I can make a separate PR with this change (as suggested by @dal24). The macro's name would then be `KOKKOS_IMPL_INVALID_INDEX`.

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:536]: Sure as long as you remove the changes you make in other pull requests here.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/src/Kokkos_Macros.hpp:536]: Done in #6641.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Move the `Impl::UnorderedMapErase` aggregate initialization debate to another PR (or explain why it belongs here)
Withdraw all the "invalid type" change form this PR and solely focus on asynchronous behavior

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp:417]: That's what we are normally doing.

**Comment by masterleinad** [comment]
> Would a proposal as follows make sense (at least the concept)? It would be the implementation of the pattern mentioned above:

I prefer just calling `fence` before and after explicitly. Given that there might be a lot of code in between the fences, I find the introduction of a wrapper more obfuscating than helpful.

**Comment by romintomasetti** [comment]
Hi @masterleinad ! I think this PR is now more focused on the topic "add `exec` argument to rehash, erase, etc".

I think I used the same style as you use *e.g.* in `Kokkos::fill_random` (when no `exec` given, first use `Kokkos::fence`, then call the overload with the default `exec`, then fence on `exec`).

**Review by masterleinad — COMMENTED** [review]
Please add/modify some tests so that the new overloads are actually tested.

File: containers/unit_tests/TestUnorderedMap.hpp [hunk 12]
```diff
@@ -158,10 +157,8 @@ struct TestErase {
   TestErase(map_type map, uint32_t num_erases, uint32_t num_duplicates)
       : m_map(map), m_num_erase(num_erases), m_num_duplicates(num_duplicates) {}
 
-  void testit() {
-    execution_space().fence();
-    Kokkos::parallel_for(m_num_erase, *this);
-    execution_space().fence();
+  void testit(const execution_space& exec) {
+    Kokkos::parallel_for(Kokkos::RangePolicy<execution_space>(exec, 0, m_num_erase), *this);
```
Comment (masterleinad) [inline containers/unit_tests/TestUnorderedMap.hpp:161]: All the other `tesit`s fence the execution space instance.
```suggestion
    Kokkos::parallel_for(Kokkos::RangePolicy<execution_space>(exec, 0, m_num_erase), *this);
    exec.fence();
```

**Review by masterleinad — COMMENTED** [review]
Have you checked that erase and rehash don't fence internally when called with execution space instances? It would be good to add a corresponding check since that seems to be the purpose of this pull request.

File: containers/src/Kokkos_Bitset.hpp [hunk 13]
```diff
@@ -130,10 +130,11 @@ class Bitset {
 
   /// set all bits to 1
   /// can only be called from the host
-  void set() {
-    Kokkos::deep_copy(m_blocks, ~0u);
+  void set(const execution_space& exec) {
```
Comment (dalg24) [inline containers/src/Kokkos_Bitset.hpp:133]: This would be a breaking change.
Same comment about reset() and clear()

**Comment by romintomasetti** [comment]
This is not a priority, and I might not even need this anymore. Closing.
